sdks/python/apache_beam/runners/worker/bundle_processor.py

```diff
@@ -198,6 +203,85 @@ def is_globally_windowed(self):
             == sideinputs._global_window_mapping_fn)
 
 
+class CombiningValueRuntimeState(userstate.RuntimeState):
+  def __init__(self, underlying_bag_state, combinefn):
+    self._combinefn = combinefn
+    self._underlying_bag_state = underlying_bag_state
+
+  def _read_accumulator(self, rewrite=True):
+    merged_accumulator = self._combinefn.merge_accumulators(
+        self._underlying_bag_state.read())
+    if rewrite:
+      self._underlying_bag_state.clear()
+      self._underlying_bag_state.add(merged_accumulator)
+    return merged_accumulator
+
+  def read(self):
+    return self._combinefn.extract_output(self._read_accumulator())
+
+  def add(self, value):
+    # Prefer blind writes, but don't let them grow unboundedly.
+    if random.random() < 0.5:
+      accumulator = self._read_accumulator(False)
+      self._underlying_bag_state.clear()
+    else:
+      accumulator = self._combinefn.create_accumulator()
+    self._underlying_bag_state.add(
+        self._combinefn.add_input(accumulator, value))
+
+  def clear(self):
+    self._underlying_bag_state.clear()
+
+
+# TODO(BEAM-5428): Implement cross-bundle state caching.
+class SynchronousBagRuntimeState(userstate.RuntimeState):
+  def __init__(self, state_handler, state_key, value_coder):
+    self._state_handler = state_handler
+    self._state_key = state_key
+    self._value_coder = value_coder
+
+  def read(self):
+    return IterableState(
+        self._state_handler, self._state_key, self._value_coder)
+
+  def add(self, value):
+    self._state_handler.blocking_append(
+        self._state_key, self._value_coder.encode(value))
+
+  def clear(self):
+    self._state_handler.blocking_clear(self._state_key)
+
+
+class UserStateContext(userstate.UserStateContext):
```

Comment:
Should we name this something else (maybe `FnApiUserStateContext`?) to differentiate it from the base?